src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs

```diff
@@ -570,25 +986,734 @@ public class ReadOnlyAttribute : System.Attribute { }
+                //         Process((ref readonly int p) => p);
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "ref readonly int p").WithArguments("System.Runtime.CompilerServices.IsReadOnlyAttribute").WithLocation(8, 18));
+        }
+
+        [Fact]
+        public void RefReadOnlyErrorsForLocalFunctionsDoNotPolluteCompilationDeclarationsDiagnostics()
+        {
+            var code = @"
+public class Test
+{
+    private int x = 0;
+    void User()
+    {
+        void local(ref readonly int x) { }
+        local(x);
+    }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(code, options: TestOptions.ReleaseModule);
+
+            compilation.DeclarationDiagnostics.Verify();
+
+            compilation.VerifyDiagnostics(
+                // (7,20): error CS0518: Predefined type 'System.Runtime.CompilerServices.IsReadOnlyAttribute' is not defined or imported
+                //         void local(ref readonly int x) { }
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "ref readonly int x").WithArguments("System.Runtime.CompilerServices.IsReadOnlyAttribute").WithLocation(7, 20));
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/19395")]
+        public void RefReadOnlyDefinitionsInsideUserDefinedIsReadOnlyAttribute_Class_NoParent()
+        {
+            // PROTOTYPE(readonlyRefs) verify bug is fixed and enable the test
+
+            var code = @"
+namespace System.Runtime.CompilerServices
+{
+    public class IsReadOnlyAttribute
+    {
+        public ref readonly int Method(ref readonly int x) => ref x;
+    }
+}";
+
+            CompileAndVerify(code, verify: false, symbolValidator: module =>
+            {
+                var isReadOnlyAttributeName = WellKnownTypes.GetMetadataName(WellKnownType.System_Runtime_CompilerServices_IsReadOnlyAttribute);
+                var type = module.ContainingAssembly.GetTypeByMetadataName(isReadOnlyAttributeName);
+
+                var method = type.GetMember<MethodSymbol>("Method");
+                Assert.Equal(type, method.GetReturnTypeAttributes().Single().AttributeClass);
+
+                var parameter = method.Parameters.Single();
+                Assert.Equal(type, parameter.GetAttributes().Single().AttributeClass);
+            });
+
+            CreateCompilationWithMscorlib(code).VerifyEmitDiagnostics(/* there should be an error */);
+            Assert.False(true, "It must produce an error about not being able to find the correct signature");
+        }
+
+        [Fact]
+        public void RefReadOnlyDefinitionsInsideUserDefinedIsReadOnlyAttribute_Class_CorrectParent()
+        {
+            var code = @"
+namespace System.Runtime.CompilerServices
+{
+    public class IsReadOnlyAttribute : System.Attribute
+    {
+        public ref readonly int Method(ref readonly int x) => ref x;
```

Comment:
Added tests for indexers, operators, properties, and overrides. There is already a test that verifies we produce errors if `IsReadOnly` was an interface, because it is lacking the constructor.